api/src/test/java/marquez/db/JobVersionDaoTest.java

```diff
@@ -0,0 +1,211 @@
+    final Optional<UUID> latestRunUuid = jobVersionDao.findLatestRunFor(jobVersionRow.getUuid());
+    assertThat(latestRunUuid).isPresent().contains(runRow.getUuid());
+  }
+
+  @Test
+  public void testUpsertJobVersionOnRunTransition() {
+    // Generate a new job meta object with an existing namespace; the namespace will also be
+    // associated with the input and output datasets for the job.
+    final JobMeta jobMeta = newJobMetaWith(NamespaceName.of(namespaceRow.getName()));
+    // (1) Add a new job; the input and output datasets for the job will also be added.
+    final JobRow jobRow =
+        DbTestUtils.newJobWith(
+            jdbiForTesting, namespaceRow.getName(), newJobName().getValue(), jobMeta);
+
+    // (2) Add a new run; the input dataset versions will also be associated with the run.
+    final RunRow runRow =
+        DbTestUtils.newRun(jdbiForTesting, jobRow.getNamespaceName(), jobRow.getName());
+
+    // Ensure the input dataset versions have been associated with the run.
+    DbTestUtils.verifyRunHasInputs(jdbiForTesting, runRow.getUuid(), jobMeta.getInputs().size());
+    // Ensure a run with the state NEW has no output dataset versions.
+    DbTestUtils.verifyRunHasOutputs(jdbiForTesting, runRow.getUuid(), 0);
+
+    // (4) Transition the run from NEW to RUNNING.
+    final Run runStarted =
+        DbTestUtils.transitionRunTo(jdbiForTesting, runRow.getUuid(), RunState.RUNNING);
+    assertThat(runStarted.getState()).isEqualTo(RunState.RUNNING);
+    assertThat(runStarted.getStartedAt()).isNotNull();
+
+    // (5) Transition the run from RUNNING to COMPLETED.
+    final Run runCompleted =
+        DbTestUtils.transitionRunWithOutputs(
+            jdbiForTesting, runRow.getUuid(), RunState.COMPLETED, jobMeta.getOutputs());
+    assertThat(runCompleted.getState()).isEqualTo(RunState.COMPLETED);
+    assertThat(runCompleted.getEndedAt()).isNotNull();
+    assertThat(runCompleted.getDurationMs()).isPresent();
+
+    // Ensure the output dataset versions have been associated with the run.
+    DbTestUtils.verifyRunHasOutputs(jdbiForTesting, runRow.getUuid(), jobMeta.getOutputs().size());
+
+    // (6) Add a new job version on the run state transition to COMPLETED.
+    final BagOfJobVersionInfo bagOfJobVersionInfo =
+        jobVersionDao.upsertJobVersionOnRunTransition(
+            jobRow.getNamespaceName(),
+            jobRow.getName(),
+            runRow.getUuid(),
+            RunState.COMPLETED,
+            newTimestamp());
+
+    // Ensure the job version is associated with the latest run.
+    DbTestUtils.verifyLatestJobVersionForRun(
+        jdbiForTesting, runRow.getUuid(), bagOfJobVersionInfo.getJobVersionRow().getUuid());
+
+    // Ensure the latest run is associated with the job version.
+    final Optional<UUID> latestRunUuid =
+        jobVersionDao.findLatestRunFor(bagOfJobVersionInfo.getJobVersionRow().getUuid());
+    assertThat(latestRunUuid).isPresent().contains(runRow.getUuid());
```

Comment:
Make the inverse of this assertion before transitioning the run to COMPLETED

Reply:
This is an assertion I wanted to add as well. But, the job version is created by calling `JobVersionDao.upsertJobVersionOnRunTransition()`. (A version is now created when done). Meaning, we won't have a reference to the version to assert the latest run UUID until the state transition. I added `JobVersion.findJobVersionFor(Run)` allowing us to assert no job version is associated with a run.